plugins/badges-backend/src/lib/BadgeBuilder/DefaultBadgeBuilder.ts

```diff
@@ -17,10 +17,16 @@
 import { InputError } from '@backstage/errors';
 import { makeBadge, Format } from 'badge-maker';
 import { BadgeBuilder, BadgeInfo, BadgeOptions, BadgeSpec } from './types';
-import { Badge, BadgeFactories } from '../../types';
+import { Badge, BadgeFactories, injectables } from '../../types';
+import { inject, injectable } from 'inversify';
 
+@injectable()
 export class DefaultBadgeBuilder implements BadgeBuilder {
-  constructor(private readonly factories: BadgeFactories) {}
+  private readonly factories: BadgeFactories;
+
+  constructor(@inject(injectables.BadgeFactories) factories: BadgeFactories) {
```

Comment:
Gotta say that having used `ApiRef`s for a while now it feels really unfortunate to have the declaration of the type here. I feel it's quite a lot nicer to have the type tied to `injectables.BadgeFactories` rather than having to declare it separately on every usage, and potentially also get that declaration wrong.

I do feel that using constructor as the public API for injection is also a bit unfortunate. It's perhaps alright for internal plugin implementations, but a fixed list of constructor arguments makes a pretty poor public API. Not sure if there's a different way you can declare this? Something like

```ts
constructor({@inject(injectables.BadgeFactories) factories}: { factories: BadgeFactories }) {
```

Even that has the problem of being a breaking change when adding known to exist dependency. I think a good property of this system would be if you can always expect a set of core implementations to exist, even in tests. Similar to how `renderInTestApp` provides a baseline for tests in the frontend where you don't need to care about certain APIs being available.

Somewhere we could end up for public APIs is a continuation of the `.from*` pattern with something like a `.fromContainer`, `.fromInversify`. Although perhaps that ends up really clunky since once you're in DI land you kinda stay there, so could be more of a `.bind(container) thing instead that just registers a factory/instance.

Just a general thing to be cautions of with all of these DI framework options is that I think they're primarily built to support development of standalone services, and not to be used as a communication layer between libraries with public API. We're going to want to end up in both of those categories, because tbh what would the point of this discussion be otherwise? a backend plugin can already use inversify internally if it wants, the benefit comes from being able to access common utilities injected by the backend, like logging, metrics, DB, etc. That is all going to need to form a stable API

Reply:
Yeah, good points. I think we have more wiggle room here to design our own API for internal and external classes when _not_ using the decorator based injection. I do think that controller injection is the way to go regardless (with an options/args object instead of explicit args), mostly because other options will end up having more downsides (setters possibly making instances half-baked, service locator possibly hiding actual dependencies since every class would have access to everything within the container).